File: hackerrank/problem solving/algorithms/2 - implementation/26 - circular array rotation/with morfo library/circular_array_rotation.go
// https://www.hackerrank.com/challenges/circular-array-rotation/problem?isFullScreen=true

package main

import (
    "fmt"
    "morfo/morfoio"
)

func main() {
    n := morfoio.Read[int]()
    nRotation := morfoio.Read[int]()
    nQueries := morfoio.Read[int]()
    arr := morfoio.ReadN[int](n)
    printQueries(arr, nRotation, nQueries)
}

// T: O(nQueries)
// S: O(1) extra space
func printQueries(arr []int, nRotation, nQueries int) {
    n := len(arr)
    offset := n - nRotation % n
    for range nQueries {
        query := morfoio.Read[int]()
        idx := (query + offset) % n
        fmt.Println(arr[idx])
    }
}

File: hackerrank/problem solving/algorithms/2 - implementation/26 - circular array rotation/with morfo library/circular_array_rotation.rs
// https://www.hackerrank.com/challenges/circular-array-rotation/problem?isFullScreen=true

use morfo::{read, read_many, read_n};

fn main() {
    let (n, n_rotation, n_queries): (usize, usize, usize) = read_many();
    let arr: Vec<i32> = read_n(n);
    print_queries(&arr, n_rotation, n_queries);
}

// T: O(n_queries)
// S: O(1) extra space
fn print_queries(arr: &[i32], n_rotation: usize, n_queries: usize) {
    let n: usize = arr.len();
    let offset: usize = n - n_rotation % n;
    for _ in 0..n_queries {
        let query: usize = read();
        let idx = (query + offset) % n;
        println!("{}", arr[idx]);
    }
}
